modules/CMakeWorkspaceCreator.pm

```diff
@@ -0,0 +1,140 @@
+package CMakeWorkspaceCreator;
+
+# ************************************************************
+# Description   : A CMake Workspace creator
+# Author        : Chad Elliott
+# Create Date   : 10/10/2022
+# ************************************************************
+
+# ************************************************************
+# Pragmas
+# ************************************************************
+
+use strict;
+use File::Basename;
+
+use CMakeProjectCreator;
+use WorkspaceCreator;
+
+use vars qw(@ISA);
+@ISA = qw(WorkspaceCreator);
+
+# ************************************************************
+# Data Section
+# ************************************************************
+
+my $version = '3.12.0';
+
+# ************************************************************
+# Subroutine Section
+# ************************************************************
+
+sub pre_workspace {
+  my($self, $fh) = @_;
+  my $crlf = $self->crlf();
+
+  $self->print_workspace_comment($fh,
+           '# CMake Workspace', $crlf,
+           '#', $crlf,
+           '# This file was generated by MPC.', $crlf,
+           '#', $crlf,
+           '# MPC Command:', $crlf,
+           '# ', $self->create_command_line_string($0, @ARGV), $crlf);
+}
+
+sub write_and_compare_file {
+  my($self, $outdir, $oname, $func, @params) = @_;
+  my $status = 1;
+  my $errorString = '';
+  my @project_dirs;
+
+  ## Set the output directory if one wasn't provided
+  $outdir = $self->get_outdir() if (!defined $outdir);
+
+  ## Remove the existing workspace, if one exists.
+  my $wsname = "$outdir/CMakeLists.txt";
+  unlink($wsname);
+
+  ## Rename the first (and hopefully the only) project in the directory to what
+  ## CMake expects.
+  my %dirs;
+  foreach my $entry ($self->sort_dependencies($self->get_projects(), 0)) {
+    my $dir = dirname($entry);
+    if (!exists $dirs{$dir}) {
+      ## Keep track of the project existing in this directory
+      $dirs{$dir} = 1;
+
+      ## Rename the project file to CMakeLists.txt and if it fails, we need to
+      ## propagate that back to the caller.
+      if (rename($entry, "$dir/CMakeLists.txt") == 0) {
+        $status = 0;
+        $errorString = "Unable to rename $entry";
+        last;
+      }
+
+      push(@project_dirs, $dir);
+    }
+    else {
+      $self->warning("Multiple projects in the same workspace are not " .
+                     "supported: $dir");
```

Comment:
Yeah, so the way that the template is built (and the way that MPC expects things to work, in general), having multiple projects in one generated CMakeLists.txt would be a bit difficult.  It probably isn't impossible, but since this is supposed to be just a takeoff point.  I was assuming that we would generate once and then hand-modify the generated files from that point on.  @mitza-oci Is that not how you envisioned it?

Reply:
If it's an deep-seated MPC assumption, then I'm okay with leaving this how this is.

Reply:
Could the project/workspace mapping work like it does for `-type make` and `-type gnuace`?  So each MPC project would generate a `ProjectName.cmake` file (containing a CMake "target" definition) and each directory that has one or more of those would have a `CMakeLists.txt` that includes `*.cmake` in the current directory (with CMake's `include`) and subdirectories (with CMake's `add_subdirectory`)?

Reply:
That is actually the way it works.  I just realized the message is incorrect.  Instead of saying "Multiple projects in the same workspace are not supported", it should say "Multiple projects in the same directory are not supported"

Reply:
Since the templates would all share the same variable names, it might get confusing or even cause incorrect compilations if we included one generated txt into another.  I guess if we made all of the project name part of the variable names, we could do this without any real issues?

Reply:
Here's more details on how I could see this working.

CMakeLists.txt only has include and add_subdirectory, like the GNUmakefile without an extension
```
include(proj1.cmake)
include(proj2.cmake)
add_subdirectory(dir1)
add_subdirectory(dir2)
````

proj1.cmake
```
add_library(proj1 src1.cpp)
target_include_directories(proj1 incdir1)
```

The definitions in proj1.cmake only impact the CMake target proj1.  I agree with what you mentioned about naming because CMake will consider this all part of the same scope like an `#include` in C++.

Reply:
Having multiple projects in a single directory is now supported.